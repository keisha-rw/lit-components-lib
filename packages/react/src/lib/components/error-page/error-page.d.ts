import * as React from 'react';
export declare const PdsErrorPage: React.FunctionComponent<Partial<{
    linkHref: String;
    layoutContainerVariant: "default" | "narrow" | undefined;
    errorCode: "403" | "404" | "500" | "503";
    linkText: String;
}> & {
    ref?: React.Ref<any> | undefined;
} & {
    style?: React.CSSProperties | undefined;
} & Partial<{}> & {
    id?: string | undefined;
    slot?: string | undefined;
    className?: string | undefined;
    children?: React.ReactNode;
}>;
