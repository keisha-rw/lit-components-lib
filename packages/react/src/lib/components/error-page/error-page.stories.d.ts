import type { Meta, StoryObj } from '@storybook/react';
import { PdsErrorPage } from './error-page';
declare const meta: Meta<typeof PdsErrorPage>;
export default meta;
type Story = StoryObj<typeof PdsErrorPage>;
export declare const Error404: Story;
export declare const Error403: Story;
export declare const Error500: Story;
export declare const Error503: Story;
