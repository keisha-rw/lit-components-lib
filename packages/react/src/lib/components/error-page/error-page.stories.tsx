import * as React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { PdsErrorPage } from './error-page';

const meta: Meta<typeof PdsErrorPage> = {
  title: 'Components/Error page',
  component: PdsErrorPage,
  tags: ['autodocs'],
};
export default meta;

type Story = StoryObj<typeof PdsErrorPage>;

export const Error404: Story = {
  render: (args: {}) => <PdsErrorPage {...args} />,
  args: {
    errorCode: '404',
    linkText: '[Insert link here]',
    linkHref: 'https://www.google.com',
  },
};
Error404.storyName = '404';

export const Error403: Story = {
  render: (args: {}) => <PdsErrorPage {...args} />,
  args: {
    errorCode: '403',
    linkText: '[Insert link here]',
    linkHref: 'https://www.google.com',
  },
};
Error403.storyName = '403';

export const Error500: Story = {
  render: (args: {}) => <PdsErrorPage {...args} />,
  args: {
    errorCode: '500',
    linkText: '[Insert link here]',
    linkHref: 'https://www.google.com',
  },
};
Error500.storyName = '500';

export const Error503: Story = {
  render: (args: {}) => <PdsErrorPage {...args} />,
  args: {
    errorCode: '503',
    linkText: '[Insert link here]',
    linkHref: 'https://www.google.com',
  },
};
Error503.storyName = '503';
