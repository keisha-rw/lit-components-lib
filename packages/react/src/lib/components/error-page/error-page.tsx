import * as React from 'react';
import { PdsErrorPage as PdsErrorPageElement } from '@keisha/design-system';
import { PdsErrorPageProps } from '../../built-component-props';
import { createPdsReactComponent } from '../../create-pds-react-component';

export const PdsErrorPage = createPdsReactComponent<PdsErrorPageProps>()({
  tagName: 'pds-error-page',
  elementClass: PdsErrorPageElement,
  react: React,
});
