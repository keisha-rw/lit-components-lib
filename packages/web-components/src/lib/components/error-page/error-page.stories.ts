import { Meta, StoryObj } from '@storybook/web-components';
import { html, nothing } from 'lit';
import './error-page';

const meta: Meta = {
  title: 'Components/Error page',
  component: 'pds-error-page',
  tags: ['autodocs'],
  argTypes: {
    linkText: { control: 'text' },
    helpHref: { control: 'text' },
    layoutContainerVariant: { control: 'text' },
  },
  render: (args) =>
    html`<pds-error-page
      errorCode="${args.errorCode || nothing}"
      linkText="${args.linkText || '[Insert link here]'}"
      linkHref="${args.linkHref || 'https://www.google.com'}"
    >
    </pds-error-page>`,
};
export default meta;

// span element is needed to wrap plain text slots https://github.com/storybookjs/storybook/issues/15963

export const Error404: StoryObj = {
  name: '404',
  args: {
    errorCode: '404',
  },
};

export const Error403: StoryObj = {
  name: '403',
  args: {
    errorCode: '403',
  },
};

export const Error500: StoryObj = {
  name: '500',
  args: {
    errorCode: '500',
  },
};

export const Error503: StoryObj = {
  name: '503',
  args: { errorCode: '503' },
};
