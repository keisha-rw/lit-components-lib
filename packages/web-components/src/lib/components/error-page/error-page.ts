import { TemplateResult, html } from 'lit';
import { localized, msg } from '@lit/localize';
import { property, state } from 'lit/decorators.js';
import { pdsCustomElement as customElement } from '../../decorators/pds-custom-element';
import { PdsElement } from '../PdsElement';
import styles from './error-page.scss?inline';
import '../layout-container/layout-container';
import '../heading/heading';
import '../text-passage/text-passage';
import '../link/link';

/**
 * @summary This component renders an error page used in simple full-page error scenarios
 */
@customElement('pds-error-page', {
  category: 'component',
  type: 'component',
  styles,
})
@localized()
export class PdsErrorPage extends PdsElement {
  connectedCallback() {
    super.connectedCallback();
    this.setLocale();
  }

  firstUpdated() {
    this.setWindowResizeHandler();
    this.getLinkSize();
  }

  /**
   * Style errorCode
   * - **403** renders the error-page used to indicate the authentication failure
   * - **404** renders the error-page used to indicate the broken link or if the page is moved
   * - **500** renders the error-page used to indicate the server is unavailable due to internal configuration issue or a temporary glitch
   * - **503** renders the error-page used to indicate if the site is unreachable, usually due to planned maintenance
   */
  @property()
  errorCode: '403' | '404' | '500' | '503' = '500';

  /**
   * Text can be passed to the link
   */
  @property()
  linkText: String;

  /**
   * A href can be passed to the link
   */
  @property()
  linkHref: String;

  /**
   * variant for layout container
   * - **narrow** renders the layout-container narrower than the default
   */
  @property({ type: String })
  layoutContainerVariant?: 'default' | 'narrow' = 'default';

  /**
   * The content that goes in the text passage within the error page
   * @internal
   */
  @state()
  textPassageContent: String;

  /**
   * The heading for the error page
   * @internal
   */
  @state()
  heading: TemplateResult;

  /**
   * The size of the link in the error page
   * @internal
   */
  @state()
  linkSize: String;

  /**
   * @internal
   */
  get classNames() {
    return {
      [this.errorCode]: !!this.errorCode,
    };
  }

  loadErrorPageContent() {
    if (this.errorCode === '404') {
      this.heading = msg(
        html`Well, that’s
          <span class="${this.classEl('styled-text')}">puzzling.</span>`,
      );
      this.textPassageContent = msg(
        'We’re sorry. It looks like the link you followed may be broken or the page may have been moved. ',
      );
    } else if (this.errorCode === '403') {
      this.heading = msg(
        html`<span class="${this.classEl('styled-text')}">Well, shoot.</span>
          Our system doesn’t recognize you.`,
      );
      this.textPassageContent = msg(
        'Sorry! It looks like you don’t have permission to access this page.',
      );
    } else if (this.errorCode === '500') {
      this.heading = msg(
        html`<span class="${this.classEl('styled-text')}">Uh-oh.</span> We’re
          experiencing a small glitch.`,
      );
      this.textPassageContent = msg(
        'Please refresh your browser or try again in a few minutes.',
      );
    } else if (this.errorCode === '503') {
      this.heading = msg(
        html`Looks like there’s a
          <span class="${this.classEl('styled-text')}">bump</span> in the road.`,
      );
      this.textPassageContent = msg(
        'Sorry! This part of our site is currently unavailable, we’re working on resolving the issue as quickly as possible.',
      );
    }
  }

  getLinkSize() {
    if (
      this.responsiveViewportSize === 'xs' ||
      this.responsiveViewportSize === 'sm'
    ) {
      this.linkSize = 'lg';
    } else {
      this.linkSize = 'xl';
    }
  }

  render() {
    return html`<div class=${this.getClass()}>
      <pds-layout-container
        variant="${this.layoutContainerVariant}"
        removePadding="all"
        class="${this.classEl('layout-container')}"
      >
        <div class="${this.classEl('container')}">
          <div class="${this.classEl('content-container')}">
            <div>
              ${this.loadErrorPageContent()}
              <pds-text-passage>
                <span class="${this.classEl('status-code')}"
                  >${msg('ERROR')} ${this.errorCode}</span
                >
                <pds-heading
                  headingTag="h1"
                  variant="display-default"
                  class="${this.classEl('heading')}"
                >
                  ${this.heading}
                </pds-heading>
                <div class="${this.classEl('text-passage')}">
                  ${this.textPassageContent}
                </div>
              </pds-text-passage>
            </div>
            <div>
              <pds-link
                variant="emphasis-inverted"
                href=${this.linkHref}
                size=${this.linkSize}
                ><span>${this.linkText}</span></pds-link
              >
            </div>
          </div>
          <div class="${this.classEl('svg-container')}">
            <div class="${this.classEl('svg')}"></div>
          </div>
        </div>
      </pds-layout-container>
    </div>`;
  }
}
