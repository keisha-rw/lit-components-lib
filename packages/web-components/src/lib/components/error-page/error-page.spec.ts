import * as path from 'path';
import initStoryshots from '@storybook/addon-storyshots';
import { assert, fixture } from '@open-wc/testing';
import { PdsErrorPage } from './error-page';
import createSnapshots from '../../../../test/utils/snapshots';

initStoryshots({
  storyKindRegex: /Error page/,
  framework: 'web-components',
  configPath: path.join(__dirname, '../../../../.storybook'),
  asyncJest: true,
  test: createSnapshots,
});

describe('ErrorPage unit tests', () => {
  let element: PdsErrorPage;

  beforeEach(async () => {
    element = await fixture(
      '<pds-error-page errorCode="404" linkText="Test" linkHref="https://www.google.com"/>',
    );
  });

  it('is an instance of PdsErrorPage', () => {
    expect(element).toBeInstanceOf(PdsErrorPage);
  });

  it('is accessible', async () => {
    await assert.isAccessible(element);
  });

  it('renders correct content for 403 errorCode', async () => {
    const element1 = await fixture(
      '<pds-error-page errorCode="403" linkText="Test" linkHref="https://www.google.com"/>',
    );
    expect(element1).toMatchSnapshot();
  });

  it('renders correct content for 500 errorCode', async () => {
    const element1 = await fixture(
      '<pds-error-page errorCode="500" linkText="Test" linkHref="https://www.google.com" />',
    );
    expect(element1).toMatchSnapshot();
  });

  it('renders correct content for 503 errorCode', async () => {
    const element1 = await fixture(
      '<pds-error-page errorCode="503" linkText="Test" linkHref="https://www.google.com"/>',
    );
    expect(element1).toMatchSnapshot();
  });

  it('Link to have correct size prop value for md/lg viewport', async () => {
    const pdslinkElement = element.shadowRoot?.querySelector(
      'pds-link',
    ) as HTMLElement;
    global.innerWidth = 800;
    global.dispatchEvent(new Event('resize'));
    expect(pdslinkElement.getAttribute('size')).toBe('xl');
  });

  it('Link to have correct size prop value for xs/sm viewport', async () => {
    element.responsiveViewportSize = 'sm';
    element.getLinkSize();
    await element.updateComplete;

    const pdslinkElement = element.shadowRoot?.querySelector(
      'pds-link',
    ) as HTMLElement;
    expect(pdslinkElement.getAttribute('size')).toBe('lg');
  });
});
