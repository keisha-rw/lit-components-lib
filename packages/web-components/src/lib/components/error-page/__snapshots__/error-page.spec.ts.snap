// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPage unit tests renders correct content for 403 errorCode 1`] = `
<pds-error-page
  errorcode="403"
  linkhref="https://www.google.com"
  linktext="Test"
/>
`;

exports[`ErrorPage unit tests renders correct content for 500 errorCode 1`] = `
<pds-error-page
  errorcode="500"
  linkhref="https://www.google.com"
  linktext="Test"
/>
`;

exports[`ErrorPage unit tests renders correct content for 503 errorCode 1`] = `
<pds-error-page
  errorcode="503"
  linkhref="https://www.google.com"
  linktext="Test"
/>
`;
